third_party/move/move-vm/runtime/src/storage/environment.rs

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    compute_code_hash,
+    config::VMConfig,
+    loader::check_natives,
+    native_functions::{NativeFunction, NativeFunctions},
+    storage::{
+        struct_name_index_map::StructNameIndexMap, ty_cache::StructInfoCache,
+        verified_module_cache::VERIFIED_MODULES_V2,
+    },
+    Module, Script,
+};
+use ambassador::delegatable_trait;
+use bytes::Bytes;
+use move_binary_format::{
+    access::{ModuleAccess, ScriptAccess},
+    errors::{Location, PartialVMError, PartialVMResult, VMResult},
+    file_format::CompiledScript,
+    CompiledModule,
+};
+use move_bytecode_verifier::dependencies;
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::{IdentStr, Identifier},
+    vm_status::{sub_status::unknown_invariant_violation::EPARANOID_FAILURE, StatusCode},
+};
+use std::sync::Arc;
+
+/// [MoveVM] runtime environment encapsulating different configurations. Shared between the VM and
+/// the code cache, possibly across multiple threads.
+pub struct RuntimeEnvironment {
+    /// Configuration for the VM. Contains information about enabled checks, verification,
+    /// deserialization, etc.
+    vm_config: VMConfig,
+    /// All registered native functions in the current context (binary). When a verified [Module]
+    /// is constructed, existing native functions are inlined in the module representation, so that
+    /// the interpreter can call them directly.
+    natives: NativeFunctions,
+
+    /// Map from struct names to indices, to save on unnecessary cloning and reduce memory
+    /// consumption. Used by all struct type creations in the VM and in code cache.
+    struct_name_index_map: StructNameIndexMap,
```

Comment:
Seems we don't need to flush this, even module publishing is failed. so we can optimize cloning `struct_name_index_map` by changing this to `Arc<StructNameIndexMap>`. 

```suggestion
    struct_name_index_map: Arc<StructNameIndexMap>,
```

Reply:
Hm... But this is true only because we will not caching anything else, so it is fine to create an index for a struct name as long as types, for instance are not cached? Indeed  👍